alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,180 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Error;
+use std::os::raw::c_void;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+use log::warn;
+
+use mio_extras::channel::{channel, Receiver};
+
+use winapi::shared::ntdef::{BOOLEAN, HANDLE, PVOID};
+use winapi::um::winbase::{RegisterWaitForSingleObject, UnregisterWait, INFINITE};
+use winapi::um::winnt::{WT_EXECUTEINWAITTHREAD, WT_EXECUTEONLYONCE};
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback for `HandleWaitSignal`, unpacks Rust closure reference and calls it
+extern "system" fn child_exit_callback<F>(ctx: PVOID, timed_out: BOOLEAN)
+where
+    F: FnOnce() + Send,
+{
+    if timed_out != 0 {
+        return;
+    }
+
+    let callback: Box<F> = unsafe { Box::from_raw(ctx as *mut F) };
+    callback();
+}
+
+/// Represents closure attached to Win32 handle wait signal
+///
+/// This allows to fire a Rust callback when subprocess exits
+pub(crate) struct HandleWaitSignal {
+    wait_handle: AtomicPtr<c_void>,
+}
+
+impl HandleWaitSignal {
+    /// Registers an asynchronous closure to call when `child_handle` wait signals.
+    ///
+    /// The `on_exit` is called on Win32 threadpool thread so it should avoid
+    /// blocking calls. See [`WT_EXECUTEINWAITTHREAD` flag docs] for details
+    ///
+    /// [`WT_EXECUTEINWAITTHREAD` flag docs]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerwaitforsingleobject
+    fn new<F>(child_handle: HANDLE, on_exit: F) -> Result<HandleWaitSignal, Error>
+    where
+        F: FnOnce() + Send,
+    {
+        let mut wait_handle: HANDLE = 0 as HANDLE;
+        let wait_notify = Box::new(on_exit);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub(crate) struct ChildProcessState {
+    _on_exit: HandleWaitSignal,
+    events: Receiver<ChildEvent>,
```

Comment:
Is there ever anything sent through this other than an `Exit` signal? If not, maybe it makes sense to alter this channel's type and name to make it more clear what it is used for?

Reply:
All of this is based on Unix implementation that converts Unix signals from child process to `ChildEvent` enum which currently has only one case. If at one point someone needs to add another `ChildEvent` case this at least marks the place in Windows implementation which is expected to receive it.

Reply:
The Unix implementation only defines the `EventedPty` interface though, right? Which is not implemented for `ChildProcessState`, but for `Pty`.

So the `ChildProcessState` is just a struct like any other that the `Pty` uses to gather the necessary informations, it's not necessarily bound by `EventedPty` and is mostly independent.

You're talking about other `ChildEvent`s, though I'm not sure what other child events we would be caring about. However since the `ChildEvent` enum itself is already structured this way with only a single field and a lot of generically named methods just for handling this single case, I don't think it's a bad idea to follow it in this PR. Though I feel like that might not have been the most efficient choice originally.

So I think it's fine to keep things mostly the way they are right now, just to make sure that our different platforms share some similarities.

Reply:
So with my limited understanding of the codebase, this could be a lot simpler for both Unix and Windows systems with `Arc<AtomicBool>` that loop probes but you still need to send an event to `mio` to wake up the IO loop so it actually checks the value of that boolean which works fine for Unix because it has out-of-the-box support for Unix signals so IO loop would know that some signal was received and would just check it. However, there is no similar support for Windows, so best bet is to push it through a channel which would wake up `mio` loop.

If that's the case, I guess you might just push `ChildEvent` from callback directly to the IO loop instead of jumping through hoops to wake up the loop to check boolean flag.

Not sure if that's the best way of doing this for both Unix and Windows but I'm not familiar enough with the codebase yet to judge that. Happy to take any suggestions.